core/src/main/java/io/undertow/server/DefaultByteBufferPool.java

```diff
@@ -148,7 +148,9 @@ public PooledByteBuffer allocate() {
                 buffer = ByteBuffer.allocate(bufferSize);
             }
         }
-        if(local != null) {
+        // limit allocationDepth runaway in scenarios where buffers flow from I/O threads to workers.
+        // Without an upper bound, allocationDepth may overflow and disable the threadLocal cache.
+        if (local != null && local.allocationDepth < 100) {
```

Comment:
Initially I had created an `allocationDepthCeiling` instance variable, and initialized it to `allocationDepth * 10`. The `100` constant could be problematic for threadlocal caches with sizes greater than 100, but I opted for this approach for simplicity.

Feel free to modify if you feel differently, I can see a case for either way.